presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java

```diff
@@ -160,37 +165,51 @@ public Page getOutput()
             checkState(process != null, "process is null");
             createTask();
             checkState(task != null, "task is null");
-            taskStatusFuture = task.start();
+            TaskInfo taskInfo = task.start();
+            if (processTaskInfo(taskInfo)) {
+                return null;
+            }
         }
 
         try {
-            if (taskStatusFuture.isDone()) {
-                // Will throw exception if the  taskStatusFuture is done with error.
-                taskStatusFuture.get();
-                Optional<TaskInfo> taskInfo = task.getTaskInfo();
-                taskInfo.ifPresent(info -> info.getTaskStatus().getFailures().forEach(e -> log.error(e.toException())));
-                Optional<SerializedPage> page = task.pollResult();
-                if (page.isPresent()) {
-                    return processResult(page.get());
-                }
-                else {
-                    if (taskInfo.isPresent() && taskInfo.get().getTaskStatus().getState().isDone()) {
-                        info.set(new NativeExecutionInfo(ImmutableList.of(taskInfo.get().getStats())));
-                        finished = true;
-                    }
-                    return null;
-                }
+            Optional<SerializedPage> page = task.pollResult();
+            if (page.isPresent()) {
+                return processResult(page.get());
             }
 
-            Optional<SerializedPage> page = task.pollResult();
-            return page.map(this::processResult).orElse(null);
+            Optional<TaskInfo> taskInfo = task.getTaskInfo();
+            if (taskInfo.isPresent() && processTaskInfo(taskInfo.get())) {
+                return null;
+            }
+
+            return null;
         }
-        catch (InterruptedException | ExecutionException e) {
+        catch (InterruptedException e) {
             log.error(e);
             throw new RuntimeException(e);
         }
     }
 
+    private boolean processTaskInfo(TaskInfo taskInfo)
+    {
+        TaskStatus taskStatus = taskInfo.getTaskStatus();
+        if (!taskStatus.getState().isDone()) {
+            return false;
+        }
+
+        if (taskStatus.getState() != TaskState.FINISHED) {
+            RuntimeException failure = taskStatus.getFailures().stream()
+                    .findFirst()
+                    .map(ExecutionFailureInfo::toException)
+                    .orElse(new PrestoException(GENERIC_INTERNAL_ERROR, "Native task failed for an unknown reason"));
+            throw failure;
+        }
+
+        info.set(new NativeExecutionInfo(ImmutableList.of(taskInfo.getStats())));
+        finished = true;
```

Comment:
Shouldn't we be setting `finished=true` even in the `!FINISHED` case above ?

Reply:
I believe 'finished' flag is only for happy path.